# Railway Incidents Data - Dataset Preparation

## This script contains the following points:
1. Importing Libraries
2. Importing Data
3. Checking Data
4. Data Wrangling
5. Data Consistency Check
6. Data Profile
7. Exporting Data

## 1. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 2. Importing Data

In [2]:
# Project folder path
path=r'C:\Users\maryn\Documents\Data Projects\Railway Incidents Data'

In [3]:
# Import dataset Rail_Equipment_Accident_Incident_Data.csv
df = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'Rail_Equipment_Accident_Incident_Data.csv'), index_col = False)

C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\1316145723.py:2: DtypeWarning: Columns (26,27,43,46,47,48,50,59,61,62,63,64,65,66,67,68,69,71,129,130,131,136) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'Rail_Equipment_Accident_Incident_Data.csv'), index_col = False)


## 3. Checking Data

In [4]:
# Display the df information 
df.head()

Reporting Railroad Code                            Reporting Railroad Name  \
0                    NICD  Northern Indiana Commuter Transportation District   
1                    NIRC  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...   
2                      CR                                            Conrail   
3                      IC                  Illinois Central Railroad Company   
4                      IC                  Illinois Central Railroad Company   

   Report Year Accident Number  \
0       2017.0        N1700013   
1       2017.0       201707188   
2       1981.0         0420001   
3       2007.0          543854   
4       2017.0          938223   

                                            PDF Link  Accident Year  \
0  https://safetydata.fra.dot.gov/Officeofsafety/...           17.0   
1  https://safetydata.fra.dot.gov/Officeofsafety/...           17.0   
2  https://safetydata.fra.dot.gov/Officeofsafety/...           81.0   
3  https://safetydata.fra.dot.gov/Officeofsafety/...            7.0   
4  https://safetydata.fra.dot.gov/Officeofsafety/...           17.0   

   Accident Month Other Railroad Code  \
0             6.0                 NaN   
1             6.0                NICD   
2             4.0                 NaN   
3             1.0                 NaN   
4            10.0                 NaN   

                                 Other Railroad Name Other Accident Number  \
0                                                NaN                   NaN   
1  Northern Indiana Commuter Transportation District              N1700013   
2                                                NaN                   NaN   
3                                                NaN                   NaN   
4                                                NaN                   NaN   

   ...  Other Railroad SMT Grouping  Other Parent Railroad Company Code  \
0  ...                          NaN                                 NaN   
1  ...            SMT-4 - CP/CN/CCD                                NICD   
2  ...                          NaN                                 NaN   
3  ...                          NaN                                 NaN   
4  ...                          NaN                                 NaN   

                  Other Parent Railroad Company Name  \
0                                                NaN   
1  Northern Indiana Commuter Transportation District   
2                                                NaN   
3                                                NaN   
4                                                NaN   

  Other Railroad Holding Company Maintenance Railroad Company Grouping  \
0                            NaN         Passenger (Formerly Commuter)   
1                   Not Assigned         Passenger (Formerly Commuter)   
2                            NaN                                   NaN   
3                            NaN                              Shippers   
4                            NaN                                   NaN   

   Maintenance Railroad Class  Maintenance Railroad SMT Grouping  \
0                     Class 3                  SMT-4 - CP/CN/CCD   
1                     Class 3                  SMT-4 - CP/CN/CCD   
2                     Class 3                                NaN   
3                     Class 3                                NaN   
4                     Class 1                  SMT-4 - CP/CN/CCD   

  Maintenance Parent Railroad Company Code  \
0                                     NIRC   
1                                     NIRC   
2                                     CRSH   
3                                     CAEZ   
4                                       CN   

            Maintenance Parent Railroad Company Name  \
0  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...   
1  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...   
2                      Consolidated Rail Corporation   
3                             

In [5]:
df.shape

(215849, 160)

In [6]:
df.columns

Index(['Reporting Railroad Code', 'Reporting Railroad Name', 'Report Year',
       'Accident Number', 'PDF Link', 'Accident Year', 'Accident Month',
       'Other Railroad Code', 'Other Railroad Name', 'Other Accident Number',
       ...
       'Other Railroad SMT Grouping', 'Other Parent Railroad Company Code',
       'Other Parent Railroad Company Name', 'Other Railroad Holding Company',
       'Maintenance Railroad Company Grouping', 'Maintenance Railroad Class',
       'Maintenance Railroad SMT Grouping',
       'Maintenance Parent Railroad Company Code',
       'Maintenance Parent Railroad Company Name',
       'Maintenance Railroad Holding Company'],
      dtype='object', length=160)

In [7]:
column_list = df.columns.tolist()
print(column_list)

['Reporting Railroad Code', 'Reporting Railroad Name', 'Report Year', 'Accident Number', 'PDF Link', 'Accident Year', 'Accident Month', 'Other Railroad Code', 'Other Railroad Name', 'Other Accident Number', 'Other Accident Year', 'Other Accident Month', 'Maintenance Railroad Code', 'Maintenance Railroad Name', 'Maintenance Accident Number', 'Maintenance Accident Year', 'Maintenance Accident Month', 'Grade Crossing ID', 'Day', 'Date', 'Time', 'Accident Type Code', 'Accident Type', 'Hazmat Cars', 'Hazmat Cars Damaged', 'Hazmat Released Cars', 'Persons Evacuated', 'Subdivision', 'Division Code', 'Division', 'Station', 'Milepost', 'State Code', 'State Abbreviation', 'State Name', 'County Code', 'County Name', 'District', 'Temperature', 'Visibility Code', 'Visibility', 'Weather Condition Code', 'Weather Condition', 'Track Type Code', 'Track Type', 'Track Name', 'Track Class', 'Track Density', 'Train Direction Code', 'Train Direction', 'Equipment Type Code', 'Equipment Type', 'Equipment Atte

## 4. Data Wrangling

### Selection columns 

In [8]:
# Selection columns 
selected_columns = [
    'Reporting Railroad Code', 'Reporting Railroad Name', 'Report Year', 'Accident Number', 
    'Accident Month', 'Accident Type', 'State Name', 'District', 'Train Speed', 'Temperature', 
    'Visibility', 'Weather Condition', 'Passengers Transported', 'Total Damage Cost', 
    'Total Persons Killed', 'Total Persons Injured', 'Latitude', 'Longitude'
]

df_selected = df[selected_columns]


In [9]:
# Check outcome
df_selected.head()

Reporting Railroad Code                            Reporting Railroad Name  \
0                    NICD  Northern Indiana Commuter Transportation District   
1                    NIRC  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...   
2                      CR                                            Conrail   
3                      IC                  Illinois Central Railroad Company   
4                      IC                  Illinois Central Railroad Company   

   Report Year Accident Number  Accident Month      Accident Type State Name  \
0       2017.0        N1700013             6.0         Derailment   ILLINOIS   
1       2017.0       201707188             6.0         Derailment   ILLINOIS   
2       1981.0         0420001             4.0     Side collision   NEW YORK   
3       2007.0          543854             1.0         Derailment  LOUISIANA   
4       2017.0          938223            10.0  Hwy-rail crossing   ILLINOIS   

   District  Train Speed  Temperature Visibility Weather Condition  \
0       4.0         10.0         65.0        Day             Clear   
1       4.0          0.0         65.0        Day             Clear   
2       1.0          0.0         28.0        Day              Snow   
3       5.0          4.0         56.0        Day            Cloudy   
4       4.0          0.0         66.0       Dark             Clear   

  Passengers Transported Total Damage Cost  Total Persons Killed  \
0                    Yes           132,013                   0.0   
1                    NaN           132,013                   0.0   
2                     No             4,235                   0.0   
3                     No             9,986                   0.0   
4                     No            13,231                   0.0   

   Total Persons Injured   Latitude  Longitude  
0                    0.0  41.884035 -87.623003  
1                    0.0  41.884035 -87.623003  
2                    0.0        NaN        NaN  
3                    0.0   0.000000   0.000000  
4                    1.0  41.054254 -87.906206

In [10]:
df_selected.shape

(215849, 18)

### Rename columns

In [11]:
# Rename columns in accordance with column name standards
df_selected.rename(columns = {'Reporting Railroad Code' : 'reporting_railroad_code'}, inplace = True)

C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\4239546074.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected.rename(columns = {'Reporting Railroad Code' : 'reporting_railroad_code'}, inplace = True)


In [12]:
# Export the selected data to avoid SettingWithCopyWarning
df_selected.to_csv(os.path.join(path, '02 Data','Prepared Data','RID_selected_data.csv'))

In [13]:
# Import the data back as a new DataFrame
df_selected_1 = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','RID_selected_data.csv'), index_col = False)

In [14]:
# Rename columns in accordance with column name standards
df_selected_1.rename(columns = {'Reporting Railroad Code' : 'railroad_code'}, inplace = True)

In [15]:
df_selected_1.rename(columns = {'Reporting Railroad Name' : 'railroad_name'}, inplace = True)

In [16]:
df_selected_1.rename(columns = {'Report Year' : 'year'}, inplace = True)

In [17]:
df_selected_1.rename(columns = {'Accident Number' : 'accident_id'}, inplace = True)

In [18]:
df_selected_1.rename(columns = {'Accident Month' : 'month'}, inplace = True)

In [19]:
df_selected_1.rename(columns = {'Accident Type' : 'accident_type'}, inplace = True)

In [20]:
df_selected_1.rename(columns = {'State Name' : 'state'}, inplace = True)

In [21]:
df_selected_1.rename(columns = {'District' : 'fra_district'}, inplace = True)

In [22]:
df_selected_1.rename(columns = {'Train Speed' : 'train_speed'}, inplace = True)

In [23]:
df_selected_1.rename(columns = {'Temperature' : 'temperature'}, inplace = True)

In [24]:
df_selected_1.rename(columns = {'Visibility' : 'visibility'}, inplace = True)

In [25]:
df_selected_1.rename(columns = {'Weather Condition' : 'weather_condition'}, inplace = True)

In [26]:
df_selected_1.rename(columns = {'Passengers Transported' : 'has_passengers'}, inplace = True)

In [27]:
df_selected_1.rename(columns = {'Total Damage Cost' : 'damage_cost'}, inplace = True)

In [28]:
df_selected_1.rename(columns = {'Total Persons Killed' : 'persons_killed'}, inplace = True)

In [29]:
df_selected_1.rename(columns = {'Total Persons Injured' : 'persons_injured'}, inplace = True)

In [30]:
df_selected_1.rename(columns = {'Latitude' : 'latitude'}, inplace = True)

In [31]:
df_selected_1.rename(columns = {'Longitude' : 'longitude'}, inplace = True)

In [32]:
# Check outcome
df_selected_1.head()

Unnamed: 0 reporting_railroad_code  \
0           0                    NICD   
1           1                    NIRC   
2           2                      CR   
3           3                      IC   
4           4                      IC   

                                       railroad_name    year accident_id  \
0  Northern Indiana Commuter Transportation District  2017.0    N1700013   
1  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...  2017.0   201707188   
2                                            Conrail  1981.0     0420001   
3                  Illinois Central Railroad Company  2007.0      543854   
4                  Illinois Central Railroad Company  2017.0      938223   

   month      accident_type      state  fra_district  train_speed  \
0    6.0         Derailment   ILLINOIS           4.0         10.0   
1    6.0         Derailment   ILLINOIS           4.0          0.0   
2    4.0     Side collision   NEW YORK           1.0          0.0   
3    1.0         Derailment  LOUISIANA           5.0          4.0   
4   10.0  Hwy-rail crossing   ILLINOIS           4.0          0.0   

   temperature visibility weather_condition has_passengers damage_cost  \
0         65.0        Day             Clear            Yes     132,013   
1         65.0        Day             Clear            NaN     132,013   
2         28.0        Day              Snow             No       4,235   
3         56.0        Day            Cloudy             No       9,986   
4         66.0       Dark             Clear             No      13,231   

   persons_killed  persons_injured   latitude  longitude  
0             0.0              0.0  41.884035 -87.623003  
1             0.0              0.0  41.884035 -87.623003  
2             0.0              0.0        NaN        NaN  
3             0.0              0.0   0.000000   0.000000  
4             0.0              1.0  41.054254 -87.906206

In [33]:
# Drop the 'Unnamed: 0' column
df_selected_1.drop(columns=['Unnamed: 0'], inplace=True)

In [34]:
# Check outcome
df_selected_1.head()

reporting_railroad_code                                      railroad_name  \
0                    NICD  Northern Indiana Commuter Transportation District   
1                    NIRC  NORTHEAST ILLINOIS REGIONAL COMMUTER RAIL CORP...   
2                      CR                                            Conrail   
3                      IC                  Illinois Central Railroad Company   
4                      IC                  Illinois Central Railroad Company   

     year accident_id  month      accident_type      state  fra_district  \
0  2017.0    N1700013    6.0         Derailment   ILLINOIS           4.0   
1  2017.0   201707188    6.0         Derailment   ILLINOIS           4.0   
2  1981.0     0420001    4.0     Side collision   NEW YORK           1.0   
3  2007.0      543854    1.0         Derailment  LOUISIANA           5.0   
4  2017.0      938223   10.0  Hwy-rail crossing   ILLINOIS           4.0   

   train_speed  temperature visibility weather_condition has_passengers  \
0         10.0         65.0        Day             Clear            Yes   
1          0.0         65.0        Day             Clear            NaN   
2          0.0         28.0        Day              Snow             No   
3          4.0         56.0        Day            Cloudy             No   
4          0.0         66.0       Dark             Clear             No   

  damage_cost  persons_killed  persons_injured   latitude  longitude  
0     132,013             0.0              0.0  41.884035 -87.623003  
1     132,013             0.0              0.0  41.884035 -87.623003  
2       4,235             0.0              0.0        NaN        NaN  
3       9,986             0.0              0.0   0.000000   0.000000  
4      13,231             0.0              1.0  41.054254 -87.906206

## 5. Data Consistency Check

### Duplicates

In [35]:
# Identify duplicate rows
duplicates = df_selected_1.duplicated()

In [36]:
# Display the duplicate rows
duplicate_rows = df_selected_1[duplicates]

print("\nDuplicate rows:")
print(duplicate_rows)


Duplicate rows:
       reporting_railroad_code                                railroad_name  \
131                         UP               Union Pacific Railroad Company   
165                        CSX                           CSX Transportation   
326                         UP               Union Pacific Railroad Company   
400                         UP               Union Pacific Railroad Company   
460                         UP               Union Pacific Railroad Company   
...                        ...                                          ...   
215688                     CSX                           CSX Transportation   
215702                      UP               Union Pacific Railroad Company   
215724                    BNSF                         BNSF Railway Company   
215726                      IC            Illinois Central Railroad Company   
215771                      CC  Chicago, Central & Pacific Railroad Company   

          year accident_id  month 

In [37]:
df_no_duplicates = df_selected_1.drop_duplicates()

In [38]:
df_no_duplicates.shape

(210672, 18)

### Missing values

In [39]:
#Looking for missing values
df_no_duplicates.isnull().sum()

reporting_railroad_code         1
railroad_name                   3
year                            1
accident_id                     0
month                           1
accident_type                   2
state                           1
fra_district                    1
train_speed                     3
temperature                     1
visibility                    238
weather_condition             392
has_passengers              10566
damage_cost                     1
persons_killed                  1
persons_injured                 1
latitude                   150788
longitude                  150772
dtype: int64

In [40]:
df_no_duplicates.dtypes

reporting_railroad_code     object
railroad_name               object
year                       float64
accident_id                 object
month                      float64
accident_type               object
state                       object
fra_district               float64
train_speed                float64
temperature                float64
visibility                  object
weather_condition           object
has_passengers              object
damage_cost                 object
persons_killed             float64
persons_injured            float64
latitude                   float64
longitude                  float64
dtype: object

### Addressing Missing Values

#### The 'latitude' and 'longitude' columns have too many missing values, making them useless for the further analysis.

In [41]:
# Drop columns 'latitude' and 'longitude'
df_no_duplicates = df_no_duplicates.drop(columns=['latitude', 'longitude'])

#### Columns 'visibility', 'weather_condition' and 'has_passengers' have object datatypes. Therefore, we replace missing values with common category 'Unknown' for these variables.

In [42]:
# Fil miising values of the variables 'visibility', 'weather_condition' and 'has_passengers' with a common category 'Unknown'
df_no_duplicates['visibility'].fillna('Unknown', inplace=True)
df_no_duplicates['weather_condition'].fillna('Unknown', inplace=True)
df_no_duplicates['has_passengers'].fillna('Unknown', inplace=True)

#### Other columns have 1 to 3 missing values, and dropping them in a data set with 210672 rows should not skew the data any more than any other way of addressing.

In [44]:
# Drop rows with any missing values
df_cleaned_na = df_no_duplicates.dropna()

In [45]:
# Ckeck the output
df_cleaned_na.shape

(210666, 16)

In [61]:
df_cleaned_na.head()

reporting_railroad_code                                      railroad_name  \
0                    NICD  Northern Indiana Commuter Transportation District   
2                      CR                                            Conrail   
3                      IC                  Illinois Central Railroad Company   
4                      IC                  Illinois Central Railroad Company   
5                      NS                   Norfolk Southern Railway Company   

     year accident_id  month      accident_type      state  fra_district  \
0  2017.0    N1700013    6.0         Derailment   ILLINOIS           4.0   
2  1981.0     0420001    4.0     Side collision   NEW YORK           1.0   
3  2007.0      543854    1.0         Derailment  LOUISIANA           5.0   
4  2017.0      938223   10.0  Hwy-rail crossing   ILLINOIS           4.0   
5  2017.0      127020   10.0         Derailment   ILLINOIS           4.0   

   train_speed  temperature visibility weather_condition has_passengers  \
0         10.0         65.0        Day             Clear            Yes   
2          0.0         28.0        Day              Snow             No   
3          4.0         56.0        Day            Cloudy             No   
4          0.0         66.0       Dark             Clear             No   
5          3.0         60.0       Dawn              Rain             No   

  damage_cost  persons_killed  persons_injured  info_missing  
0     132,013             0.0              0.0         False  
2       4,235             0.0              0.0         False  
3       9,986             0.0              0.0         False  
4      13,231             0.0              1.0         False  
5     148,289             0.0              0.0         False

### Mixed data types

In [46]:
# Check for mixed types
for col in df_cleaned_na.columns.tolist():
    weird = (df_cleaned_na[[col]].map(type)!=df_cleaned_na[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_cleaned_na[weird])>0:
        print (col)

In [47]:
# Check actual data types
df_cleaned_na.dtypes

reporting_railroad_code     object
railroad_name               object
year                       float64
accident_id                 object
month                      float64
accident_type               object
state                       object
fra_district               float64
train_speed                float64
temperature                float64
visibility                  object
weather_condition           object
has_passengers              object
damage_cost                 object
persons_killed             float64
persons_injured            float64
dtype: object

In [48]:
# Convert 'year', 'month', and 'fra_district' columns to integer
df_cleaned_na['year'] = df_cleaned_na['year'].astype(int)
df_cleaned_na['month'] = df_cleaned_na['month'].astype(int)
df_cleaned_na['fra_district'] = df_cleaned_na['fra_district'].astype(int)

C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\2870443088.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_na['year'] = df_cleaned_na['year'].astype(int)
C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\2870443088.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_na['month'] = df_cleaned_na['month'].astype(int)
C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\2870443088.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .l

In [49]:
# Convert 'year' column to integers using .loc
df_cleaned_na.loc[:, 'year'] = pd.to_numeric(df_cleaned_na['year'], errors='coerce').fillna(0).astype(int)

# Convert 'month' column to integers using .loc
df_cleaned_na.loc[:, 'month'] = pd.to_numeric(df_cleaned_na['month'], errors='coerce').fillna(0).astype(int)

# Convert 'fra_district' column to integers using .loc
df_cleaned_na.loc[:, 'fra_district'] = pd.to_numeric(df_cleaned_na['fra_district'], errors='coerce').fillna(0).astype(int)

In [50]:
df_cleaned_na.loc[:, 'damage_cost'] = pd.to_numeric(df_cleaned_na['damage_cost'], errors='coerce').fillna(0).astype(float)

In [51]:
# Check the outcome
df_cleaned_na.dtypes

reporting_railroad_code     object
railroad_name               object
year                         int32
accident_id                 object
month                        int32
accident_type               object
state                       object
fra_district                 int32
train_speed                float64
temperature                float64
visibility                  object
weather_condition           object
has_passengers              object
damage_cost                 object
persons_killed             float64
persons_injured            float64
dtype: object

In [52]:
# Identify non-numeric values
non_numeric = df_cleaned_na[~df_cleaned_na['damage_cost'].apply(lambda x: str(x).replace('.', '', 1).isdigit())]
print("Non-numeric values in 'damage_cost' column:")
print(non_numeric)

Non-numeric values in 'damage_cost' column:
Empty DataFrame
Columns: [reporting_railroad_code, railroad_name, year, accident_id, month, accident_type, state, fra_district, train_speed, temperature, visibility, weather_condition, has_passengers, damage_cost, persons_killed, persons_injured]
Index: []


In [53]:
# Convert 'damage_cost' to numeric, filling NaNs with 0 if necessary
df_cleaned_na['damage_cost'] = pd.to_numeric(df_cleaned_na['damage_cost'], errors='coerce')

C:\Users\maryn\AppData\Local\Temp\ipykernel_8528\1432201088.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_na['damage_cost'] = pd.to_numeric(df_cleaned_na['damage_cost'], errors='coerce')


In [54]:
# Using .loc to ensure setting the value on the original DataFrame
df_cleaned_na.loc[:, 'damage_cost'] = pd.to_numeric(df_cleaned_na['damage_cost'], errors='coerce')

In [55]:
# Fill NaNs with 0 if needed
df_cleaned_na.loc[:, 'damage_cost'] = df_cleaned_na['damage_cost'].fillna(0).astype(float)

In [56]:
# Check the outcome
df_cleaned_na.dtypes

reporting_railroad_code     object
railroad_name               object
year                         int32
accident_id                 object
month                        int32
accident_type               object
state                       object
fra_district                 int32
train_speed                float64
temperature                float64
visibility                  object
weather_condition           object
has_passengers              object
damage_cost                float64
persons_killed             float64
persons_injured            float64
dtype: object

# 6. Data Profile

In [57]:
# Summary statistics
df_cleaned_na.describe().astype(int)

year   month  fra_district  train_speed  temperature  damage_cost  \
count  210666  210666        210666       210666       210666       210666   
mean     1991       6             4           11           56            0   
std        14       3             1           15           23           14   
min      1975       1             1            0          -65            0   
25%      1979       3             3            3           40            0   
50%      1989       6             4            5           59            0   
75%      2004       9             6           12           74            0   
max      2022      12             8          150          862          978   

       persons_killed  persons_injured  
count          210666           210666  
mean                0                0  
std                 0                3  
min                 0                0  
25%                 0                0  
50%                 0                0  
75%                 0                0  
max                47              558

In [58]:
# Value counts for 'temperature' 
df_cleaned_na['temperature'].value_counts(dropna = False)

temperature
 70.0     11761
 60.0     11104
 50.0     10605
 65.0      9309
 40.0      9172
          ...  
 754.0        1
 862.0        1
 841.0        1
-33.0         1
 264.0        1
Name: count, Length: 180, dtype: int64

In [59]:
# Calculate the value counts and sort them from lowest value to highest
temperature_counts_sorted = df_cleaned_na['temperature'].value_counts(dropna=False).sort_index()

In [60]:
# Check the output
print(temperature_counts_sorted)

temperature
-65.0     1
-60.0     2
-49.0     1
-42.0     2
-40.0     7
         ..
 754.0    1
 841.0    1
 842.0    2
 852.0    1
 862.0    1
Name: count, Length: 180, dtype: int64


In [63]:
# Identifying Outliers
outliers = df_cleaned_na[df_cleaned_na['temperature'] > 134]
outliers.shape[0]

22

In [64]:
# Calculate the mean of the non-outlier values
mean_temperature = df_cleaned_na[df_cleaned_na['temperature'] <= 134]['temperature'].mean()

In [65]:
# Replace outliers with the mean
df_cleaned_na.loc[df_cleaned_na['temperature'] > 134, 'temperature'] = mean_temperature

In [66]:
# Check the output
temperature_counts_sorted = df_cleaned_na['temperature'].value_counts(dropna=False).sort_index()
print(temperature_counts_sorted)

temperature
-65.0     1
-60.0     2
-49.0     1
-42.0     2
-40.0     7
         ..
 114.0    4
 115.0    8
 116.0    5
 117.0    1
 118.0    2
Name: count, Length: 160, dtype: int64


In [67]:
# Summary statistics after changes made
df_cleaned_na.describe().astype(int)

year   month  fra_district  train_speed  temperature  damage_cost  \
count  210666  210666        210666       210666       210666       210666   
mean     1991       6             4           11           55            0   
std        14       3             1           15           22           14   
min      1975       1             1            0          -65            0   
25%      1979       3             3            3           40            0   
50%      1989       6             4            5           59            0   
75%      2004       9             6           12           74            0   
max      2022      12             8          150          118          978   

       persons_killed  persons_injured  
count          210666           210666  
mean                0                0  
std                 0                3  
min                 0                0  
25%                 0                0  
50%                 0                0  
75%                 0                0  
max                47              558

In [68]:
# Check unique values of categorical data
for column in df_cleaned_na.select_dtypes(include='object').columns:
    unique_values = df_cleaned_na[column].unique()
    print(f"\nUnique values in {column}:")
    print(unique_values)


Unique values in reporting_railroad_code:
['NICD' 'NIRC' 'CR' 'IC' 'NS' 'UP' 'IHB' 'BNSF' 'CSX' 'GRW' 'BPRR' 'GTW'
 'EJE' 'WC' 'FCRD' 'ATK' 'BRC' 'KYLE' 'WTA' 'PAL' 'NOPB' 'NOGC' 'KCS' 'PW'
 'GRS' 'MBTA' 'CRSH' 'PGR' 'TCWR' 'MNNR' 'CP' 'PATH' 'NJTR' 'SNJX' 'SW'
 'MNCW' 'LI' 'IORY' 'CWRO' 'WE' 'MCCP' 'SLWC' 'KRR' 'GNBC' 'CORP' 'PNWR'
 'SEPA' 'URR' 'RJCP' 'ESPN' 'DL' 'LC' 'SCRF' 'RCPE' 'WTNN' 'TNTX' 'RVSC'
 'PTRA' 'PMSW' 'AWRR' 'GVSR' 'CMC' 'DGNO' 'TXPF' 'AGCR' 'TXR' 'FWWR'
 'CMTY' 'ATR' 'WTLC' 'RLAX' 'UTAX' 'USR' 'BB' 'NPB' 'SV' 'NECR' 'WACR'
 'TMBL' 'CWA' 'WER' 'WSOR' 'KNWA' 'WW' 'AO' 'ABS' 'MNBR' 'TASD' 'ATN'
 'AKMD' 'LRWN' 'AZCR' 'SMRT' 'SJVR' 'SCAX' 'PHL' 'PCMZ' 'NCTC' 'ARZC'
 'LAJ' 'ITSL' 'SERA' 'SCRT' 'TRC' 'SFRV' 'CFRC' 'FEC' 'CCT' 'TSWS' 'SFBR'
 'MET' 'ACEX' 'RPRC' 'SDNX' 'NSWS' 'GOOS' 'VCRR' 'ATAX' 'AZER' 'CBRY'
 'SPSR' 'ALM' 'MNA' 'ARS' 'NLA' 'AM' 'LRPA' 'LNW' 'DQE' 'EDW' 'AGR' 'OUCH'
 'EARY' 'BHRR' 'CCHA' 'ALE' 'TNHR' 'COEH' 'LXVR' 'SOO' 'PCC' 'KFR' 'SBVR'
 'SCR' 'PSAP' 'EWG

In [69]:
# Mode and cardinality for each categorical column
modes = df_cleaned_na.mode().iloc[0]
cardinality = df_cleaned_na.nunique()
print("\nMode and Cardinality:")
for column in df_cleaned_na.select_dtypes(include='object').columns:
    print(f"{column}: Mode = {modes[column]}, Cardinality = {cardinality[column]}")


Mode and Cardinality:
reporting_railroad_code: Mode = UP, Cardinality = 990
railroad_name: Mode = Union Pacific Railroad Company, Cardinality = 987
accident_id: Mode = 001, Cardinality = 166889
accident_type: Mode = Derailment, Cardinality = 13
state: Mode = ILLINOIS, Cardinality = 50
visibility: Mode = Day, Cardinality = 5
weather_condition: Mode = Clear, Cardinality = 7
has_passengers: Mode = No, Cardinality = 3


## 7. Exporting Data

In [71]:
# Exporting df_cleaned_na as a RID_clean in csv format
df_cleaned_na.to_csv(os.path.join(path, '02 Data','Prepared Data', 'RID_clean_1.csv'))